internal/provider/vm/vm_affinity_rule_schema.go

```diff
@@ -60,135 +59,99 @@ func vmAffinityRuleSchema() superschema.Schema {
 			"name": superschema.StringAttribute{
-				Resource: &schemaR.StringAttribute{
-					Required:            true,
+				Common: &schemaR.StringAttribute{
 					MarkdownDescription: "VM affinity rule name.",
 				},
+				Resource: &schemaR.StringAttribute{
+					Required: true,
+				},
+				DataSource: &schemaD.StringAttribute{
+					MarkdownDescription: " Required if `id` is not set.",
+					Optional:            true,
+					Validators: []validator.String{
+						stringvalidator.ExactlyOneOf(path.MatchRoot("name"), path.MatchRoot("id")),
+					},
+				},
 			},
 			"polarity": superschema.StringAttribute{
-				Resource: &schemaR.StringAttribute{
-					Required:            true,
+				Common: &schemaR.StringAttribute{
 					MarkdownDescription: fmt.Sprintf("One of `%s`, `%s`", govcdtypes.PolarityAffinity, govcdtypes.PolarityAntiAffinity),
+				},
+				Resource: &schemaR.StringAttribute{
+					Required: true,
 					PlanModifiers: []planmodifier.String{
 						stringplanmodifier.RequiresReplace(),
 					},
 					Validators: []validator.String{
 						stringvalidator.OneOf(govcdtypes.PolarityAffinity, govcdtypes.PolarityAntiAffinity),
 					},
 				},
+				DataSource: &schemaD.StringAttribute{
+					Optional: true,
+				},
 			},
 			"enabled": superschema.BoolAttribute{
-				Resource: &schemaR.BoolAttribute{
-					Optional:            true,
-					Computed:            true,
-					Default:             booldefault.StaticBool(true),
+				Common: &schemaR.BoolAttribute{
 					MarkdownDescription: "`True` if this affinity rule is enabled. Default is `true`.",
 				},
-			},
-			"required": superschema.BoolAttribute{
 				Resource: &schemaR.BoolAttribute{
 					Optional: true,
 					Computed: true,
 					Default:  booldefault.StaticBool(true),
+				},
+				DataSource: &schemaD.BoolAttribute{
+					Optional: true,
+				},
+			},
+			"required": superschema.BoolAttribute{
+				Common: &schemaR.BoolAttribute{
 					MarkdownDescription: "`True` if this affinity rule is required. When a rule is mandatory, " +
 						"a host failover will not power on the VM if doing so would violate the rule. Default is `true`.",
 				},
+				Resource: &schemaR.BoolAttribute{
+					Optional: true,
+					Computed: true,
+					Default:  booldefault.StaticBool(true),
+				},
+				DataSource: &schemaD.BoolAttribute{
+					Optional: true,
+				},
 			},
 			"vm_ids": superschema.ListAttribute{
-				Resource: &schemaR.ListAttribute{
-					Required:            true,
-					ElementType:         types.StringType,
+				Common: &schemaR.ListAttribute{
 					MarkdownDescription: "List of VM IDs to apply the affinity rule to.",
+					ElementType:         types.StringType,
+				},
+				Resource: &schemaR.ListAttribute{
+					Required: true,
 					Validators: []validator.List{
 						listvalidator.SizeAtMost(2),
 						listvalidator.ValueStringsAre(fstringvalidator.IsURN()),
 					},
 				},
+				DataSource: &schemaD.ListAttribute{
+					Optional: true,
+				},
 			},
```

Comment:
```suggestion
				Common: &schemaR.ListAttribute{
					MarkdownDescription: "List of VM IDs",
					ElementType:         types.StringType,
				},
				Resource: &schemaR.ListAttribute{
					MarkdownDescription: " to apply the affinity rule to.",
					Required: true,
					Validators: []validator.List{
						listvalidator.SizeAtMost(2),
						listvalidator.ValueStringsAre(fstringvalidator.IsURN()),
					},
				},
				DataSource: &schemaD.ListAttribute{
					MarkdownDescription: " associated to the affinity rule.",
					Optional: true,
				},
			},
```

This attribute is realy `Optional` in datasource ?